src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -358,28 +358,41 @@ private void WriteAsPaxSharedInternal(Span<byte> buffer, out long actualLength)
             _checksum = WriteChecksum(tmpChecksum, buffer);
         }
 
-        // All formats save in the name byte array only the ASCII bytes that fit.
+        // All formats save in the name byte array only the UTF8 bytes that fit.
         private int WriteName(Span<byte> buffer)
         {
-            ReadOnlySpan<char> src = _name.AsSpan(0, Math.Min(_name.Length, FieldLengths.Name));
-            Span<byte> dest = buffer.Slice(FieldLocations.Name, FieldLengths.Name);
-            int encoded = Encoding.ASCII.GetBytes(src, dest);
-            return Checksum(dest.Slice(0, encoded));
+            ReadOnlySpan<char> name = _name;
+            int utf16NameTruncatedLength = GetUtf16TruncatedTextLength(name, FieldLengths.Name);
+
+            ReadOnlySpan<char> truncatedName = name.Slice(0, utf16NameTruncatedLength);
+            Span<byte> destination = buffer.Slice(FieldLocations.Name, FieldLengths.Name);
+            Encoding.UTF8.GetBytes(truncatedName, destination);
+
+            return Checksum(destination);
```

Comment:
This is doing a checksum over the whole destination, even if the whole thing isn't used.  Is that correct?

Reply:
The checksum is just adding all bytes values which are zeroed at the beginning. So unused bytes wont impact the result.

Reply:
But they will take time to compute.

Reply:
For some reason I thought `Slice`ing was worse than iterating the whole `destination` but I guess is not?

Reply:
Unless you're generally having to iterate the whole thing, it's not. Plus, it looks like a bug to checksum garbage.

Reply:
If this is a bug, it appears that there is no test coverage for the checksum.

Reply:
@GSPP this was not a bug, `destination` was zeroed at the callsite, therefore passing the whole buffer to `Checksum` was no different to passing a `Slice` of it.
FWIW: I addressed this suggestion to avoid iterating the whole buffer, as Stephen suggested.